### ISSUE TITLE
Add support for publisher policy

### ISSUE THREAD

**Issue by yunhaoling** [Azure/azure-sdk-for-python#5315]
@annatisch commented on [Tue Dec 18 2018](https://github.com/Azure/azure-event-hubs-python/issues/87)

Publisher Policy overview for reference:
https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-features#publisher-policy

One of his references is this blog post:
https://blogs.msdn.microsoft.com/servicebus/2015/02/02/event-hub-publisher-policy-in-action/

That blog, along with everything else he and I can find all reference the need to create the token through this method in .NET:
```
String token = SharedAccessSignatureTokenProvider.GetPublisherSharedAccessSignature(parameters)
```

**Comment by alanjcastonguay** [comment]
What was the conclusion here?

**Comment by yunhaoling** [comment]
Hello @alanjcastonguay , this feature isn't in our current plan.

Are you waiting on this feature implementation? I'd like to hear more about your user case/thoughts on this feature, I can bring it to the sdk team for discussion. If it's good/necessary to have this feature in the python sdk, then we'll make a plan.

Feel free to re-open the issue.

**Comment by alanjcastonguay** [comment]
I had hoped to use Publisher Policies similar to an API Key, crafting one for each of our customer IoT aggregator/sender IoT devices (so maybe 1-20 per customer), with each Publisher Name identifying a customer. And that the customer's device would use that when sending a stream of IoT metrics to our EventHub instance. And on receipt (eg in Capture or from an EH Consumer), all events would be marked with a field holding the Publisher Name which sent the event.

**Comment by yunhaoling** [comment]
Hi @alanjcastonguay ,

I'm introducing @annatisch to this conversation as she may knows better than me on this.

Hi Anna, I'm not quire sure how this feature is supposed to work -- When receiving events which are sent with Publisher Policies, would the events be marked with a field holding the publisher name as Alan described?

Or is it something can be achieved by manually setting the `properties`/`application_properties` of `EventData`?

**Comment by alanjcastonguay** [comment]
(I must admit I’m not absolutely certain what I described is the intent. I’m piecing together this belief from the dotnet examples.)

**Comment by alanjcastonguay** [comment]
Of note I was hoping the receiver would get the publisher’s identity (and be able to differentiate events by it) without the publisher being able to impersonate another publisher (eg by trivially writing the publisher ID in the event body).

**Comment by yunhaoling** [comment]
Hi @alanjcastonguay , sorry for the late responding, I just came back from a vacation.

Can you share me the link to the dotnet samples related to this PublisherPolicy?

--------

> Of note I was hoping the receiver would get the publisher’s identity (and be able to differentiate events by it) without the publisher being able to impersonate another publisher (eg by trivially writing the publisher ID in the event body).

After reading the docs in https://docs.microsoft.com/en-us/archive/blogs/servicebus/event-hub-publisher-policy-in-action, I see that the client side still need to define a publisher ID locally like.
```string publisherId = "1";``` and use this value in other places.

With that saying, you still need to define publisher id by yourself.
You can achieve this by manually setting the [`properties` of an EventData](https://github.com/Azure/azure-sdk-for-python/blob/0a05f49b2aa13730a501885c3723f69052b70f3c/sdk/eventhub/azure-eventhub/azure/eventhub/_common.py#L199)

```
# for sending
event = EventData("body")
event.properties = { "publisher_id" : "1" }

# for receiver
publisher_id = received_event.properties.get("publisher_id")
```

Please correct me if I'm wrong.

**Comment by alanjcastonguay** [comment]
@yunhaoling I have not managed to re-locate the dotnet examples I mentioned earlier - I should have linked them then.

https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-features#publisher-policy suggests that the Publisher ID is part of both the Connection String/URL and the SAS Token. The holders of those SAS Tokens are authorized to publish to an EventHub, and those SAS Tokens can be revoked. I don't think the property is supposed to be written (impersonated/spoofed) by the sender.

> You don't have to create publisher names ahead of time, but they must match the SAS token used when publishing an event, in order to ensure independent publisher identities.

I think https://docs.microsoft.com/en-us/azure/event-hubs/authenticate-shared-access-signature#authenticating-event-hubs-publishers-with-sas supports this view and refers to https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.sharedaccesssignaturetokenprovider.getpublishersharedaccesssignature?view=azure-dotnet, but doesn't explicitly state the url where that SAS token would be used.


https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-features#publisher-policy contains a url structure `//[my namespace].servicebus.windows.net/[event hub name]/publishers/[my publisher name]`, and something similar appears in https://github.com/VladVin/SleetMonitor/blob/00d359b339cee8833108a17c5ead3e91907f6414/SleetMonitor/app/src/main/java/vladvin/sleetmonitor/data_proc/EventHubSender.java#L25 and in the C sdk a similar url fragment is built at https://github.com/Azure/azure-event-hubs-c/blob/ddf4512c36805cfa64688c6d85447172f70f96e2/eventhub_client/src/eventhubauth.c#L710

In 2016, @clemensv wrote https://github.com/Azure/azure-event-hubs/commit/b633366dd11f38ec44ef9272ef842c58bb62ab2c which _removes_ descriptive text about Publisher Policies "until client adds support".

Azure/azure-event-hubs-java#233 is trying to do something similar to me, but in Java. In that issue @SreeramGarlapati describes a limitation of publisher-id forcing an (sometimes undesirable) partition-key, and future work being needed before solidifying client libraries. It's unclear whether that future work was done. https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-features#publisher-policy also says "When using publisher policies, the PartitionKey value is set to the publisher name. To work properly, these values must match.". @SreeramGarlapati can you weigh in here?

I haven't found any example python code sending to an EventHub URL containing `/publishers/[my publisher name]`. It's really hard to search for that, though.

**Comment by YijunXieMS** [comment]
Hi @alanjcastonguay, 
A SAS token has expiry time, so in your use case, do the customer IoT aggregator / sender devices need to provide a new SAS token with a new expiry time before it expires while they keep sending events? How often do they send events and what do you expect them to do when the token expires?

**Comment by yunhaoling** [comment]
- The publisher policy could be achieved by manually constructing the connection string or eventhub name to include the publisher name:

```python
from azure.eventhub import EventHubProducerClient
CONNECTION_STR = os.environ['EVENT_HUB_CONN_STR']
EVENTHUB_NAME = os.environ['EVENT_HUB_NAME'] + "/publishers/samplepublisher"
producer = EventHubProducerClient.from_connection_string(
    conn_str=CONNECTION_STR,
    eventhub_name=EVENTHUB_NAME
)
#or
from azure.identity import DefaultAzureCredential
FULLY_QUALIFIED_NAMESPACE = os.environ['EVENT_HUB_HOSTNAME']
EVENTHUB_NAME = os.environ['EVENT_HUB_NAME'] + "/publishers/samplepublisher"
credential = DefaultAzureCredential()
producer_client = EventHubProducerClient(
    fully_qualified_namespace=FULLY_QUALIFIED_NAMESPACE,
    eventhub_name=EVENTHUB_NAME,
    credential=credential
)
```

- to control the lifecycle of the token, please refer to the sample here: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/eventhub/azure-eventhub/samples/async_samples/authenticate_with_sas_token_async.py#L30-L37

**Comment by vjrantal** [comment]
If someone is looking for a functional end-to-end sample in Python, you can find one here https://github.com/vjrantal/event-hub-publisher-policy-sample.
